src/main/java/org/opensearch/knn/jni/FaissService.java

```diff
@@ -28,16 +40,67 @@
  *      src/main/java/org/opensearch/knn/common/KNNConstants.java
  */
 class FaissService {
+    private static Logger logger = LogManager.getLogger(FaissService.class);
 
     static {
         AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
-            System.loadLibrary(KNNConstants.FAISS_JNI_LIBRARY_NAME);
+
+            // Even if the underlying system supports AVX2, users can override and disable it by using the
+            // 'knn.faiss.avx2.disabled' setting by setting it to true in the opensearch.yml configuration
+            if (!isFaissAVX2Disabled() && isAVX2SupportedBySystem()) {
+                System.loadLibrary(KNNConstants.FAISS_AVX2_JNI_LIBRARY_NAME);
+            } else {
+                System.loadLibrary(KNNConstants.FAISS_JNI_LIBRARY_NAME);
+            }
+
             initLibrary();
             KNNEngine.FAISS.setInitialized(true);
             return null;
         });
     }
 
+    /**
+     * Verify if the underlying system supports AVX2 SIMD Optimization or not
+     * 1. If the architecture is aarch64 return false.
+     * 2. If the operating system is windows return false.
+     * 3. If the operating system is macOS, use oshi-core library to verify if the cpu flags
+     *    contains 'avx2' and return true if it exists else false.
+     * 4. If the operating system is linux, read the '/proc/cpuinfo' file path and verify if
+     *    the flags contains 'avx2' and return true if it exists else false.
+     */
+    private static boolean isAVX2SupportedBySystem() {
+        if (Platform.isARM() || Platform.isWindows()) {
+            return false;
+        }
+
+        if (Platform.isMac()) {
+            try {
+                return AccessController.doPrivileged((PrivilegedExceptionAction<Boolean>) () -> {
+                    String flags = SysctlUtil.sysctl("machdep.cpu.leaf7_features", "empty");
+                    return (flags.toLowerCase(Locale.ROOT)).contains("avx2");
+                });
+            } catch (Exception e) {
+                logger.error("[KNN] Error fetching cpu flags info. [{}]", e.getMessage());
+                e.printStackTrace();
```

Comment:
This outputs to stderr and shouldn't be used in production code (except maybe in test classes).  If you want the stack trace in the logs, add the exception to `logger.error()` at the end.

Reply:
makes sense. Updated it